### PR TITLE
Add clouds/meshes file format converter

### PR DESCRIPTION
**Pull request by VictorLamoine**
Aims to replace:
```
convert_pcd_ascii_binary.cpp
pcd2ply.cpp
obj2pcd.cpp
obj2ply.cpp
obj2vtk.cpp
pcd2ply.cpp
pcd2vtk.cpp
ply2pcd.cpp
ply2vtk.cpp
vtk2obj.cpp
vtk2pcd.cpp
vtk2ply.cpp
```

This will also add STL format converter.

### CODE REVIEW

**Comment by VictorLamoine** [comment]
A first review would be appreciated, the program is working fine though I have not extensively tested it.

File: io/tools/convert_clouds_meshes.cpp [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+ */
+int
+main (int argc,
+      char** argv)
+{
+  // Display help
+  if (pcl::console::find_switch (argc, argv, "-h") != 0 || pcl::console::find_switch (argc, argv, "--help") != 0)
+  {
+    displayHelp (argc, argv);
+    return (-1);
```
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: I would return 0 here.

File: io/tools/convert_clouds_meshes.cpp [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    PCL_INFO("Loaded a point cloud with %d points (total size is %d) and the following channels:\n%s\n",
+             cloud->width * cloud->height, cloud->data.size (), pcl::getFieldsList (*cloud).c_str ());
+
+    if (!savePointCloud (cloud, argv[file_args[1]], output_type))
+    {
+      PCL_ERROR("Failed to save %s.\n", argv[file_args[0]]);
+      return (-1);
+    }
+  }
+
```
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: Why are you writing a special case for pcd files and not following the same logic of trying a mesh first and then switching to point cloud?

File: io/tools/convert_clouds_meshes.cpp [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    if (saveMesh (mesh, argv[file_args[1]], output_type))
+      return (-1);
+  }
+  else if (boost::filesystem::path (argv[file_args[0]]).extension () != ".stl")  // STL files cannot be point clouds
+  {
+    // OBJ, PLY or VTK
+    PCL_WARN("Could not load %s as a mesh, trying as a point cloud instead.\n", argv[file_args[1]]);
+    pcl::PCLPointCloud2::Ptr cloud (new pcl::PCLPointCloud2);
+
```
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: I think this is missing the load function equivalent to what you done in [L269](https://github.com/PointCloudLibrary/pcl/pull/1442/files#diff-de31078c18e04439315cfa29b613433aR269), ```pcl::io::load```

File: io/tools/convert_clouds_meshes.cpp (hunk 1 above)
Comment (VictorLamoine) [inline io/tools/convert_clouds_meshes.cpp]: Agreed

File: io/tools/convert_clouds_meshes.cpp (hunk 2 above)
Comment (VictorLamoine) [inline io/tools/convert_clouds_meshes.cpp]: I did it like this first because I thought about the special parameters for PCD but I think I'll remove this now for simplicity.

File: io/tools/convert_clouds_meshes.cpp (hunk 3 above)
Comment (VictorLamoine) [inline io/tools/convert_clouds_meshes.cpp]: Not finished, I forgot to add a comment :smile: 
Thanks for the review!

File: io/tools/convert_clouds_meshes.cpp [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+ * @return (true) on success, false otherwise.
+ */
+bool
+savePointCloud (pcl::PCLPointCloud2::Ptr input,
+                std::string output_file,
+                int output_type)
+{
+  if (boost::filesystem::path (output_file).extension () == ".obj")
+  {
+    //TODO: Support precision
```
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: If .obj needs to be saved as a mesh, why not invoke ```saveMesh()```, analogous to what you do in [L175](https://github.com/PointCloudLibrary/pcl/pull/1442/files#diff-de31078c18e04439315cfa29b613433aR175). The same seems to be applicable for [L131](https://github.com/PointCloudLibrary/pcl/pull/1442/files#diff-de31078c18e04439315cfa29b613433aR131) and [L136](https://github.com/PointCloudLibrary/pcl/pull/1442/files#diff-de31078c18e04439315cfa29b613433aR136).

File: io/tools/convert_clouds_meshes.cpp [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+    PCL_INFO("Loaded a point cloud with %d points (total size is %d) and the following channels:\n%s\n",
+             cloud->width * cloud->height, cloud->data.size (), pcl::getFieldsList (*cloud).c_str ());
+
+    if (!savePointCloud (cloud, argv[file_args[1]], output_type))
+    {
+      PCL_ERROR("Failed to save %s.\n", argv[file_args[0]]);
+      return (-1);
+    }
+  }
```
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: In case your .stl file is corrupted and cannot be loaded as a mesh, the program will exit normally.

**Comment by SergioRAgostinho** [comment]
Looking forward to see this merged :)

File: io/tools/convert_clouds_meshes.cpp (hunk 4 above)
Comment (VictorLamoine) [inline io/tools/convert_clouds_meshes.cpp]: That's right let's not duplicate code!

It's not possible for the STL though, STL file format does not allow to store a point cloud, so if you have only a point cloud you can't save an STL file at all.

**Comment by VictorLamoine** [comment]
I think I fixed everything above, the program is now "finished".
It's hard to test everything, so much combinations!

Color is lost during conversion into an OBJ file, thats the only issue I have identified right now.

File: io/tools/convert_clouds_meshes.cpp [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+savePointCloud (pcl::PCLPointCloud2::Ptr input,
+                std::string output_file,
+                int output_type)
+{
+  if (boost::filesystem::path (output_file).extension () == ".obj")
+  {
+    //TODO: Support precision
+    //FIXME: Color is lost during conversion (OBJ supports color)
+    if (output_type == BINARY || output_type == BINARY_COMPRESSED)
+      PCL_WARN ("OBJ file format supports only ASCII.\n");
```
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: This warning will also be printed in ```saveMesh()``` better to just remove it. This branch of your if statement can pretty much disappear and be moved to the else branch with ply and vtk.

Comment (VictorLamoine) [inline io/tools/convert_clouds_meshes.cpp]: True, I fixed that

**Comment by VictorLamoine** [comment]
I've just added an option to transform a mesh into a point cloud using the `-c` or `--cloud` switches. It just discards all faces in the file.

**Comment by SergioRAgostinho** [comment]
Sorry, I wasn't clear enough on my last comment about the warning. Note that [L102-110](https://github.com/VictorLamoine/pcl/commit/d4997f7d31361630111299939d47732492898dc3#diff-de31078c18e04439315cfa29b613433aR102) are similar to [L139-144](https://github.com/VictorLamoine/pcl/commit/d4997f7d31361630111299939d47732492898dc3#diff-de31078c18e04439315cfa29b613433aR139), which basically means you can delete L102-110.

**Comment by VictorLamoine** [comment]
Oh that's right I must have not read the right lines when reading your comment :smile: 
All of these remarks makes the code much cleaner!

File: io/tools/convert_clouds_meshes.cpp (hunk 4 above)
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: Really minor comment, notice that you had a similar situation on saving PCD as a mesh and for that one you emit a warning and force the saving as a point cloud. 

My idea was to do the same for an attempt to save a STL as a point cloud. Give the warning and force saving it as a mesh.

File: io/tools/convert_clouds_meshes.cpp (hunk 6 above)
Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: Looked at the [source](https://github.com/PointCloudLibrary/pcl/blob/master/io/src/vtk_lib_io.cpp#L78) and it definitely looks like bug, especially given the fact that you can load OBJ files in the function above :D 

If you patched it you could even move L155-165 into L174-188 adding the additional check for the ```output_type```.

Comment (SergioRAgostinho) [inline io/tools/convert_clouds_meshes.cpp]: Forget about my last suggestion. I just noticed that there are seemingly close functionalities implemented in ```pcl::io::save``` and ```pcl::io::savePolygonFile```. This should probably be addressed in another pull request.  Thanks for you effort!

**Comment by VictorLamoine** [comment]
@jspricke  Friendly pi(n)g :pig:

File: io/tools/convert_clouds_meshes.cpp [hunk 7]
```diff
@@ -0,0 +1,314 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2009, Willow Garage, Inc.
```
Comment (jspricke) [inline io/tools/convert_clouds_meshes.cpp]: Can you update the copyright?

File: io/tools/convert_clouds_meshes.cpp [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+  }
+  else if (boost::filesystem::path (output_file).extension () == ".stl")
+  {
+    PCL_ERROR ("STL file format does not support point clouds! Aborting.\n");
+    return (false);
+  }
+  else  // OBJ, PLY and VTK
+  {
+    //TODO: Support precision
+    //FIXME: Color is lost during OBJ conversion (OBJ supports color)
```
Comment (jspricke) [inline io/tools/convert_clouds_meshes.cpp]: Wouldn't that be rather a bug in the conversion function?

Comment (VictorLamoine) [inline io/tools/convert_clouds_meshes.cpp]: It's completely missing! See #1472

File: io/tools/convert_clouds_meshes.cpp [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+
+  // Convert parsed output type to output type
+  int output_type (BINARY);
+  if (!parsed_output_type.empty ())
+  {
+    if (parsed_output_type == "ascii")
+      output_type = ASCII;
+    else if (parsed_output_type == "binary_compressed")
+      output_type = BINARY_COMPRESSED;
+    else
```
Comment (taketwo) [inline io/tools/convert_clouds_meshes.cpp]: Why "binary" is not handled here?

File: io/tools/convert_clouds_meshes.cpp [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    file_args.push_back (pcd_files[i]);
+  for (int i = 0; i < ply_files.size (); ++i)
+    file_args.push_back (ply_files[i]);
+  for (int i = 0; i < stl_files.size (); ++i)
+    file_args.push_back (stl_files[i]);
+  for (int i = 0; i < vtk_files.size (); ++i)
+    file_args.push_back (vtk_files[i]);
+  // Re-order, the first argument is always the input file
+  std::sort (file_args.begin (), file_args.end ());
+  if (file_args.size () != 2)
```
Comment (taketwo) [inline io/tools/convert_clouds_meshes.cpp]: This was already tested before in line 223.

**Comment by taketwo** [comment]
Victor, great work!

I find the filename argument handling to be unnecessary verbose though. Instead of lines 213-217, 223-228, 247-265 we can write something like (untested):

```cpp
const std::vector<std::string> supported_extensions = {".obj", ".pcd", ".ply", ".stl", ".vtk"};
std::vector<int> filename_args;
for (int i = 1; i < argc; ++i)
  for (size_t j = 0; j < supported_extensions.size (); ++j)
    if (boost::algorithm::ends_with (argv[i], supported_extensions[j])
    {
      filename_args.push_back (i);
      break;
    }

if (filename_args.size () != 2)
  // error message, return

std::string input = filename_args[0];
std::string output = filename_args[1];
```

**Comment by VictorLamoine** [comment]
I've done it your way Sergey, it's shorter and still very readable. I did not use the C++11 vector initializer because C++11 is not enabled in PCL.

**Comment by taketwo** [comment]
>  I did not use the C++11 vector initializer because C++11 is not enabled in PCL.

Yeah, right, I suspected that line would not compile with the old standard :)

What about the name of the tool, maybe just `pcl_converter` or `pcl_convert`?

**Comment by VictorLamoine** [comment]
Now it's named `pcl_converter`

**Comment by taketwo** [comment]
Thanks! I haven't tried it, but the code looks good for me.

Do we want to merge this before the release or immediately after?

**Comment by VictorLamoine** [comment]
I don't mind; I think it's a useful tool so I would merge before.
One more thing to add to your giant changelog :)

**Comment by SergioRAgostinho** [comment]
Agree with Victor on both statements. Kudos for your massive changelog Sergey ;)

**Comment by taketwo** [comment]
Thanks guys.
@jspricke please merge if you think it's ready.
